prog/TC2MEI-js.hs

```
@@ -0,0 +1,77 @@
+
+module Main where
+
+import Data.ByteString.Char8
+import Options.Applicative
+import TabCode.Serialiser.MEIXML.Converter
+import TabCode.Serialiser.MEIXML.Serialiser (meiDoc)
+import TabCode.Options
+import TabCode.Parser (parseTabcode)
+import Text.XML.Generator
+
+import GHCJS.Marshal(fromJSVal)
+import GHCJS.Foreign.Callback (Callback, syncCallback1')
+import Data.JSString (JSString, pack, unpack)
+import GHCJS.Types (JSRef, JSVal, jsval)
+import JavaScript.Object (Object, create, setProp)
+import Data.Maybe (fromJust)
+
+-- blog: http://weblog.luite.com/wordpress/?p=14
+-- atom: http://edsko.net/2015/02/14/atom-haskell/
+-- https://www.reddit.com/r/haskell/comments/59jy6h/how_to_use_haskell_as_a_library_from_js_via_ghcjs/
+-- parser tutorial: https://www.futurelearn.com/courses/functional-programming-haskell/0/steps/27222
+
+#ifdef __GHCJS__
+foreign import javascript unsafe
+  "console.log($1);"
+  consoleLog :: JSString -> IO ()
+
+foreign import javascript unsafe
+  "tc2mei = $1"
+  set_callback :: Callback (JSVal -> IO JSVal) -> IO ()
+#else
+consoleLog = error "consoleLog: only available from JavaScript"
+set_callback = error "set_callback: only available from JavaScript"
+#endif
+
+main :: IO ()
+main = do
+    callback <- syncCallback1' $ \jv -> do
+      (str :: String) <- Data.JSString.unpack . fromJust <$> fromJSVal jv
+      let mei = tabcodeToMeiXml str
+      (o :: Object) <- create
+      setProp (Data.JSString.pack "mei" :: JSString) (jsval . Data.JSString.pack $ mei) o
+      return $ jsval o
+    set_callback callback
+
+tabcodeToMeiXml :: String -> String
+tabcodeToMeiXml tabcodeStr =
+  Data.ByteString.Char8.unpack $ xrender $ doc defaultDocInfo $ meiDoc meiXml (xmlIds options)
+  where
+    options = TCOptions
+      { parseMode = Permissive
+      , structure = BarLines
+      , xmlIds = WithXmlIds
+      }
+    Right tabcode = parseTabcode options tabcodeStr
```

By pattern matching only against the `Right` constructors of these two `Either` types you're ignoring any errors that may come up in either parsing the input string or converting the tabcode into MEI XML. Under GHC, this kind of pattern match failure at runtime results in the program crashing. I'm not sure what the behaviour would be under GHCJS but I'm guessing it won't be pretty.

You should consider what you want to do in the case of invalid input tabcode (easily possible) and in the case of an MEI serialisation error (less likely). One simple technique for dealing with the two possibilities would be to use a `case` expression.